### DISCUSSION TITLE
Using `notifications/tools/list_changed`

### DISCUSSION THREAD

**Discussion by evacchi** [modelcontextprotocol/.github#76] [Q&A] [answered]
### Pre-submission Checklist

- [X] I have verified that this discussion would not be more appropriate as an issue in a specific repository
- [X] I have searched existing discussions to avoid duplicates

### Discussion Topic

I have been trying to find an example usage for the `tools/list_changed` API.

From my understanding, a server should declare the capability:

```json
      "capabilities": {
        "resources": {},
        "tools": {
          "listChanged": true
        }
      }
```

then, when the list changes, it should send the message, upon which the client (I guess) would send a new `tools/list` request. 

Using Claude Desktop as a client, I slightly modified the [weather server](https://modelcontextprotocol.io/docs/first-server/typescript) to send a few notifications and append a new tool to the list, but it does not seem like the change is followed by a `tools/list` request from the client.

Is Claude Desktop implementing this feature, and if it is, what am I missing?

Thanks!

**Comment by jspahrsummers** [comment, marked as answer]
You're using it correctly, but Claude Desktop doesn't support this at the moment—sorry for the confusion.

**Reply by evacchi** [reply]
thanks!

**Reply by MaXal** [reply]
@jspahrsummers Are there any estimates when this will be available? Our MCP connects to IDE and it's weird that you need to restart Claude Client after starting IDE.

**Reply by astromechza** [reply]
I've been struggling to find _any_ MCP clients that respond to this notification. Tried Claude, Cline, Dive, etc.

**Reply by Nexlen** [reply]
The MCP Inspector also doesn't seem to support this

**Reply by johnjjung** [reply]
https://github.com/modelcontextprotocol/typescript-sdk/blob/592c91fc86607a7c11ec60da5a31e58396448cea/src/server/index.ts#L167

The typescript SDK seems to have it but it looks like it's missing implementation

**Reply by johnjjung** [reply]
https://github.com/modelcontextprotocol/typescript-sdk/pull/239

Added this PR 🤞

**Reply by b0o** [reply]
I feel like I'm going crazy - about 5-ish days ago, I tried adding/removing tools using the TypeScript SDK and issuing the `list_changed` notification, and Claude Desktop updated its list of tools accordingly without needing to refresh the page. Then, a day or two later, it stopped working.

I'm wondering if this functionality was temporarily added and then removed, or maybe feature flag gated, or maybe I'm just hallucinating the whole thing??

**Reply by johnjjung** [reply]
This functionality doesn't exist in the typescript sdk yet so you're probably experiencing the client doing a tool refresh not because the server pubbed a notification.

I got comments back so I think I can get this merged in the next commit just need to find time.

**Reply by dprophete** [reply]
any updates on this ? Using claude version 0.9.2, and an MCP server with typescript sdk 1.10.1, it seems it is still not working.

**Comment by strowk** [comment]
You can use Github Copilot. It supports list changed notifications (I tested it with my mcp hot reloading tool for MCP called synf)

**Comment by younaman** [comment]
> ```json
>  "resources": {},
> ```
Hi @jspahrsummers,

I have a question regarding MCP's discovery mechanism.

The MCP document states that when the list of tools changes, a message will be sent to the client, which should then send a new tools/list request. However, I'm wondering about what happens if I only modify the function body or return value of a tool, without changing any of the metadata (such as the name, parameters, or description).

From what I’ve observed, MCP seems to only send a notification when metadata changes. So my question is:

If I hot-update the internal logic of a tool, but leave the metadata unchanged, will MCP still send a notification to the client? Or will it ignore this kind of update?

Thank you in advance! Looking forward to hearing your insights.

**Reply by olaservo** [reply]
Looking at the spec and the Typescript SDK at least, my understanding is that a notification should only sent when the contract changes, including the name, description, parameters, etc.  Here is is a concrete example of the `_createRegisteredTool` method in [this file](https://github.com/modelcontextprotocol/typescript-sdk/blob/main/src/server/mcp.ts#L650), where the `update` function includes this behavior:

```
update: (updates) => {
  if (typeof updates.name !== "undefined" && updates.name !== name) {
    delete this._registeredTools[name]
    if (updates.name) this._registeredTools[updates.name] = registeredTool
  }
  if (typeof updates.description !== "undefined") registeredTool.description = updates.description
  if (typeof updates.paramsSchema !== "undefined") registeredTool.inputSchema = z.object(updates.paramsSchema)
  if (typeof updates.callback !== "undefined") registeredTool.callback = updates.callback
  if (typeof updates.annotations !== "undefined") registeredTool.annotations = updates.annotations
  if (typeof updates.enabled !== "undefined") registeredTool.enabled = updates.enabled
  this.sendToolListChanged()
}
```

I think the intention of the notification is that the client/LLM will be notified about any changes that could affect how the tool should be used.  So if the internal logic changes in a way that the LLM should know about, then I think the assumption is that the description should change too.

**Reply by younaman** [reply]
Thanks! I will make a deep review:)

**Comment by conikeec** [comment]
@jspahrsummers any updates on this issue with Claude Desktop? Please respond

**Comment by tadasant** [comment]
Closing as this issue is about Claude Desktop, not MCP. I encourage interested parties to use [Anthropic's support channels](https://support.anthropic.com/en/articles/9015913-how-to-get-support) for information related to their products.
